### ISSUE TITLE
Tests often hang after #3122

### ISSUE THREAD

**Issue by japgolly** [scala-js/scala-js#3175]
I built a local copy of 0.6.22-SNAPSHOT to give #3122 a spin. I tested it on my main project which has 37 modules, 10 of which build with Scala.JS.

Running tests, I often saw failures and then SBT hanging, seemingly unable to recover from the failures. In one case, the failure was simply

```
[error] PhantomJS has crashed. Please read the bug reporting guide at
[error] <http://phantomjs.org/bug-reporting.html> and file a bug report.
```
followed by SBT hanging after the other modules' tests complete.

In another instance the failure looked to be in ScalaJS itself:
```
java.lang.IllegalStateException: Unknown opcode: 5

  phantomjs://code/phantomjs-launcher9060106815877187942.js:9 in onError


  phantomjs://code/phantomjs-launcher9060106815877187942.js:11 in onError
  file:///tmp/shipreq-sjs-22.sbt/webapp-client-project/target/scala-2.12/webapp-client-project-test-fastopt.js:26478 (in function "handleMessage__T__V")

  phantomjs://code/phantomjs-launcher9060106815877187942.js:13
  file:///tmp/shipreq-sjs-22.sbt/webapp-client-project/target/scala-2.12/webapp-client-project-test-fastopt.js:181656

  phantomjs://code/phantomjs-launcher9060106815877187942.js:13
  file:///tmp/1510141923375-0/comSetup.js:30

  phantomjs://code/phantomjs-launcher9060106815877187942.js:13
```
followed, again, by hanging.

I ran jstack and uploaded the output here:
https://www.dropbox.com/sh/ouqmuqouqtvoivv/AAAgDaapef7LVPHhJyx6LWgca?dl=0

**Comment by gzm0** [comment]
Thank you for testing this!

I should have seen this coming. Two problems:
- New RPC interface does not fail all requests when runner terminates.
- Exception handling in the execution context is not good enough (needs to terminate the run).

**Comment by gzm0** [comment]
I have opened #3175 that should fix the hanging. It does not contain anything for the "unknown opcode" issue, but I suspect that this might be a cascading failure: opcodes can be dynamically registered on both ends, so if something get's out of sync because one side fails, this is likely to occur.

**Comment by gzm0** [comment]
@japgolly There is a fix for this in 0.6. If you could try it out, that'd be great.

**Comment by japgolly** [comment]
No worries, let me know and I'll test it out

**Comment by sjrd** [comment]
Now the fix has really been merged in the 0.6.x branch. ;)

**Comment by japgolly** [comment]
Ok so I tried it out. I ran my tests a bunch of times using 0.6.19, 0.6.21 and master.

In terms of speed, master is now on par with 0.6.19 :+1: whilst 0.6.21 as we already know is slower.

In terms of robustness, master mostly worked flawlessly. There was only one issue which was the `Unknown opcode: 5` error occurred again. It occurred on the first run, and then over the next 5 runs it didn't occur again.
